lib/cretonne/src/ir/frontend.rs

```diff
@@ -0,0 +1,764 @@
+/// - the last instruction of each block is a terminator instruction which has no natural sucessor,
+///   and those instructions can only appear at the end of extended blocks.
+///
+/// The parameters of Cretonne IL instructions are Cretonne IL values, which can only be created
+/// as results of other Cretonne IL instructions. To be able to create variables redefined multiple
+/// times in your program, use the `def_var` and `use_var` command, that will maintain the
+/// correspondance between your variables and Cretonne IL SSA values.
+///
+/// The first block for which you call `switch_to_block` will be assumed to be the beginning of
+/// the function.
+///
+/// At creation, a `FunctionBuilder` instance borrows an already allocated `Function` which it
+/// modifies with the information stored in the mutable borrowed
+/// [`ILBuilder`](struct.ILBuilder.html). The function passed in argument should be newly created
+///  with [`Function::with_name_signature()`](../function/struct.Function.html), whereas the
+/// `ILBuilder` can be kept as is between two function translations.
+///
+/// # Errors
+///
+/// The functions below will panic in debug mode whenever you try to modify the Cretonne IL
+/// function in a way that violate the coherence of the code. For instance: switching to a new
+/// `Ebb` when you haven't filled the current one with a terminator instruction, inserting a
+/// return instruction with arguments that don't match the function's signature.
+impl<'a, Variable> FunctionBuilder<'a, Variable>
+    where Variable: EntityRef + Hash + Default
+{
+    /// Creates a new FunctionBuilder structure that will operate on a `Function` using a
+    /// `IlBuilder`.
+    pub fn new(func: &'a mut Function,
+               builder: &'a mut ILBuilder<Variable>)
+               -> FunctionBuilder<'a, Variable> {
+        builder.clear();
+        FunctionBuilder {
+            func: func,
+            builder: builder,
+            position: Position {
+                ebb: Ebb::default(),
+                basic_block: Block::default(),
+                pristine: true,
+            },
+            pristine: true,
+        }
+    }
+
+    /// Creates a new `Ebb` for the function and returns its reference.
+    pub fn create_ebb(&mut self) -> Ebb {
+        let ebb = self.func.dfg.make_ebb();
+        self.builder.ssa.declare_ebb_header_block(ebb);
+        *self.builder.ebbs.ensure(ebb) = EbbData {
+            filled: false,
+            sealed: false,
+        };
+        ebb
+    }
+
+    /// After the call to this function, new instructions will be inserted into the designed
+    /// block, in the order they are declared.
+    ///
+    /// When inserting the terminator instruction (which doesn't have a falltrough to its immediate
+    /// successor), the block will be declared filled and it will not be possible to append
+    /// instructions to it.
+    pub fn switch_to_block(&mut self, ebb: Ebb) {
```

Comment:
I think we have to take the point of view of the user calling the API here: for him, switching to the first block should not be different from swtiching to any other block. The `self.pristine` and `fill_function_args_values` are just implementation quirks, and in my opinion they should be hidden from the user. This is also subjective though.